### ISSUE TITLE
Add an option to limit disk and RAM usage by scheduler queue

### ISSUE THREAD

**Issue by Prometheus3375** [scrapy/scrapy#6085]
## Summary

An option to limit RAM and disk usage by the scheduler queue will make the engine to take new requests from the spiders only if there is available space.

## Motivation

Recently we've run into the issue of high disk usage by the scheduler queue. We are going through company registry and making a lot of requests. These requests are the only requests spider makes. Sample code:

```python
    # crn - company registration number
    def start(self, /) -> Iterator:
        # Reset generators to start from 0 after restarts
        for generator in self.generators:
            generator.reset()

        it = ((gen, crn) for gen in self.generators for crn in gen)
        for generator, crn in it:
            self.requests_in_progress += 1
            yield Request(
                self.url_pattern_check_crn.format(crn),
                self.check_crn_status,
                dont_filter=True,
                errback=self.check_crn_status_failed,
                cb_kwargs=dict(crn=crn, generator=generator),
                )
```
Currently, generators produce 142,685,210 unique CRNs. Requests can end with 404 (company not found) or 200 (company found).

After reaching ~110k successful requests disk queue occupies 10GB. Meanwhile, RAM usage does not exceed 200MB.

## Describe alternatives you've considered

Currently, we resolved the issue with increasing `CONCURRENT_REQUESTS` and `CONCURRENT_REQUESTS_PER_DOMAIN` to big enough number, but this helps only if the site can afford so many connections or does not have any rate limiting techniques.

## Additional context

There is `SCRAPER_SLOT_MAX_ACTIVE_SIZE` setting which is described as follows:

> Soft limit (in bytes) for response data being processed.
> While the sum of the sizes of all responses being processed is above this value, Scrapy does not process new requests.

"Scrapy does not process new requests" means Scrapy does not take new requests from the spider or does not put already scheduled requests to the downloader?

We are also using FIFO queue for this spider, but I do not think this matters.

**Comment by Gallaecio** [comment]
Somewhat related to https://github.com/scrapy/scrapy/pull/3237, i.e. future work on that direction may help here.

**Comment by wRAR** [comment]
> "Scrapy does not process new requests" means Scrapy does not take new requests from the spider or does not put already scheduled requests to the downloader?

The code (`self._needs_backout()` which calls `self.scraper.slot.needs_backout()`) is in https://github.com/scrapy/scrapy/blob/5b0b00212e44f0a6ddef339ab0263cc22997b527/scrapy/core/engine.py#L164

It is about taking the next request from the scheduler and putting it into the downloader (and also about processing `start_requests()`).

**Comment by bhavuk2002** [comment]
I am willing to work on this issue pls assign me.
Only thing is i need little bit guidance.

**Comment by Gallaecio** [comment]
I don’t think this is a good choice for a first contribution to Scrapy.

**Comment by bhavuk2002** [comment]
Any other suggestions?

**Comment by Gallaecio** [comment]
https://github.com/scrapy/scrapy/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22

Anything else is best to assume to not be good first issues, although some might be.

**Comment by Gallaecio** [comment]
I am thinking we should close this issue.

Given how increasing concurrency was the workaround that worked here, I think this issue should be solved through request priority.

The issue cannot be that all those requests come from `start_requests`, since `start_requests` are only fed when the scheduler is empty at the moment, lazily. So if all requests came from `start_requests`, this issue would not happen.

So I have to assume the issue comes from crawling, from the callbacks of some requests yielding additional requests.

Under that assumption, assigning a higher priority to requests that cannot yield additional requests from their callback than to requests that can yield additional requests should solve the issue.

And if so, I see no need for any change in Scrapy about this. This kind of optimization requires a knowledge of your spider callbacks and the target website, so I don’t think there are many “automatic” improvements for these scenarios that we can implement in Scrapy itself, I really think the request priority feature is the best solution here.

@Prometheus3375 Could you confirm if my assumption is correct, or if not, provide additional details?

**Comment by Prometheus3375** [comment]
In the particular example I described in OP, all requests come from one callback (`start`) which is called once the single starting request is processed. Callbacks from those requests do not generate any new requests, only items. In that case it is not possible to add priority to some requests as they all have the same origin.

On the other note that was the time when we used `JOBDIR`, hence disk usage. Later we set `JOBDIR` to `None`, but given that was more than year ago, I do not recall any issues with RAM usage.

Basically, the issue was in the fact that the scheduler consumed many requests from `start`, saved them to disk, but due to low concurrency the engine was unable to process all of them in time, but the scheduler ignored this fact and continued to consume more requests from `start`.

In any case, it would be nice to have any kind of control over the number of scheduled requests, either via RAM/disk limit or via number of total requests scheduled at a time.

**Comment by Gallaecio** [comment]
It sounds like this would be solved by moving those requests to `start_requests`, from which requests are only read when the scheduler is empty and the number of ongoing requests is below the configured concurrency.

If using a callback was a workaround to be able to load those requests from asynchronous code (for a single URL a blocking `requests` call in `start_requests` would have been fine), I think the proper solution would be https://github.com/scrapy/scrapy/pull/4467 and maybe we should close this as a duplicate of https://github.com/scrapy/scrapy/issues/5627.
